File: src/main/java/kg/manasuniversity/usbtypec/manashelper/service/ObisClient.java
package kg.manasuniversity.usbtypec.manashelper.service;

import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.client.RestTemplate;

@Service
public class ObisClient {
  private final RestTemplate restTemplate;

  public ObisClient(RestTemplate restTemplate) {
    this.restTemplate = restTemplate;
  }

  public String fetchLoginPageHtml() {
    ResponseEntity<String> response = restTemplate
            .exchange("https://obistest.manas.edu.kg/site/login", HttpMethod.GET, null, String.class);
    return response.getBody();
  }

  public void sendLoginRequest(String studentNumber, String plainPassword, String csrfToken) {


    HttpHeaders headers = new HttpHeaders();
    headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);

    MultiValueMap<String, String> formData = new LinkedMultiValueMap<>();
    formData.add("LoginForm[username]", studentNumber);
    // Even though the field is named password_hash, it expects the plain password
    formData.add("LoginForm[password_hash]", plainPassword);
    formData.add("_csrf", csrfToken);

    HttpEntity<MultiValueMap<String, String>> loginEntity = new HttpEntity<>(formData, headers);

    ResponseEntity<String> response = restTemplate
            .exchange("https://obistest.manas.edu.kg/site/login", HttpMethod.POST, loginEntity, String.class);
  }

  public String fetchAttendancePageHtml() {
    ResponseEntity<String> response = restTemplate
            .exchange("https://obistest.manas.edu.kg/vs-ders/taken-lessons", HttpMethod.GET, null, String.class);
    return response.getBody();
  }

  public String fetchExamsPageHtml() {
    ResponseEntity<String> response = restTemplate
            .exchange("https://obistest.manas.edu.kg/vs-ders/taken-grades", HttpMethod.GET, null, String.class);
    return response.getBody();
  }
}

File: src/manashelper/scraping/obis_client.py
import httpx

from manashelper.scraping.obis_parser import parse_login_page_csrf_token

OBIS_BASE_URL = "https://obistest.manas.edu.kg"
OBIS_LOGIN_PATH = "/site/login"
OBIS_ATTENDANCE_PATH = "/vs-ders/taken-lessons"
OBIS_EXAM_GRADES_PATH = "/vs-ders/taken-grades"


class ObisLoginError(Exception):
    pass


def _new_http_client() -> httpx.AsyncClient:
    return httpx.AsyncClient(base_url=OBIS_BASE_URL, follow_redirects=True, timeout=15.0)


async def _login(client: httpx.AsyncClient, student_number: str, plain_password: str) -> None:
    login_page = await client.get(OBIS_LOGIN_PATH)
    login_page.raise_for_status()
    csrf_token = parse_login_page_csrf_token(login_page.text)

    response = await client.post(
        OBIS_LOGIN_PATH,
        data={
            "LoginForm[username]": student_number,
            "LoginForm[password_hash]": plain_password,
            "_csrf": csrf_token,
        },
    )
    response.raise_for_status()
    if response.url.path.rstrip("/") == OBIS_LOGIN_PATH:
        raise ObisLoginError("Invalid OBIS credentials")


class ObisClient:
    async def verify_credentials(self, student_number: str, plain_password: str) -> None:
        async with _new_http_client() as client:
            await _login(client, student_number, plain_password)

    async def fetch_attendance_html(self, student_number: str, plain_password: str) -> str:
        async with _new_http_client() as client:
            await _login(client, student_number, plain_password)
            response = await client.get(OBIS_ATTENDANCE_PATH)
            response.raise_for_status()
            return response.text

    async def fetch_exam_grades_html(self, student_number: str, plain_password: str) -> str:
        async with _new_http_client() as client:
            await _login(client, student_number, plain_password)
            response = await client.get(OBIS_EXAM_GRADES_PATH)
            response.raise_for_status()
            return response.text
